akka-projection-core/src/main/scala/akka/projection/internal/SourceProviderAdapter.scala

```diff
@@ -36,6 +36,12 @@ import akka.stream.scaladsl.Source
   }
 
   def extractOffset(envelope: Envelope): Offset = delegate.extractOffset(envelope)
+}
+
+@InternalApi private[akka] class VerifiableSourceProviderAdapter[Offset, Envelope](
+    delegate: javadsl.VerifiableSourceProvider[Offset, Envelope])
+    extends SourceProviderAdapter[Offset, Envelope](delegate)
+    with scaladsl.VerifiableSourceProvider[Offset, Envelope] {
 
   override def verifyOffset(offset: Offset): OffsetVerification = delegate.verifyOffset(offset)
 }
```

Comment:
I could use some help here.  

I'm not sure how to use the same adapter pattern for the difference javadsl `SourceProvider` variants.  Now that there are several combinations of providers available I need to dynamically adapt the javadsl version on an as needed basis.  I could move these concerns into the Projection implementations themselves, but that seems like it would add a lot of ceremony to the implementations.  I could make all `SourceProviders` implement the required scala and java dsl implementations, but this would mean repeating a lot of what happens in `SourceProviderAdapter` across provider implementations.

Reply:
/cc @patriknw @renatocaval @ignasi35

Reply:
One option is to not use inheritance, but treat it as capabilities. 

```scala
trait SourceProvider[Offset, Envelope]
trait VerifiableSourceProvider[Offset, Envelope]

class MySourceProvider[Long, Env] 
	extends SourceProvider[Long, Env] with VerifiableSourceProvider[Long, Env]
```
then 

```scala
object VerifiableSourceProvider {
	def verify(srcProvider:SourceProvider, offset) = {
		srcProvider match {
      case _: VerifiableSourceProvider =>  // do verification
      case _ => true
    }
  }
}
```

whenever we need it we use the impl in its companion object. 

I don't think we need to have a javadsl/scaladsl variant because the verify method in `VerifiableSourceProvider` is just a predicate. We can have only one trait. 

The VerifiableSourceProvider companion will need two `verify` variants though.

Reply:
@seglo Why is this VerifiableSourceProviderAdapter needed at all?
Can't the KafkaSourceProviderImpl implement both scaladsl and javadsl, as we have in master? https://github.com/akka/akka-projection/blob/master/akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderImpl.scala#L45

Reply:
> Why is this VerifiableSourceProviderAdapter needed at all?

You're right, it's not.  It was just a leftover from experimentation last week before I asked for advice.